File: ruby/etsm/test/ab.rb
require '../lib/etsm.rb'

class Foo
    def initialize
        @sm = Etsm::StateMachine.new(self)
        @a = Etsm::State.new(self.method(:EnterA), self.method(:ExitA))
        @b = Etsm::State.new(self.method(:EnterB), nil)
        @output = ''
    end

    def EnterA()
        @output += ' ->A '
    end

    def ExitA()
        @output += ' A-> '
    end

    def EnterB()
        @output += ' ->B '
    end

    # no exit for B
    # def ExitB()
    #     @output += ' B-> '
    # end

    def Test()
        raise "state != nil" unless @sm.IsIn(nil)
        @sm.Transition(@a)
        raise "state != A" unless @sm.IsIn(@a)
        @sm.Transition(@b)
        raise "state != B" unless @sm.IsIn(@b)
        @sm.Transition(nil)
        raise "state != nil" unless @sm.IsIn(nil)
        raise 'Output != " ->A  A->  ->B "' unless @output == ' ->A  A->  ->B '
    end
end

foo = Foo.new
foo.Test()

File: node/etsm/test/ab.js
var etsm = require('./../etsm');
var assert = require('assert');

class Foo {
    constructor() {
        this.sm = new etsm.StateMachine(this);
        this.a = new etsm.State(this.EnterA, this.ExitA);
        this.b = new etsm.State(this.EnterB, null);
        this.output = '';
    }

    EnterA(){
        this.output += ' ->A ';
        assert(this.sm.Transition(this.a) == false);
    }

    ExitA(){
        this.output += ' A-> ';
        assert(this.sm.Transition(this.a) == false);
    }

    EnterB(){
        this.output += ' ->B ';
        assert(this.sm.Transition(this.a) == false);
    }

    // don't implement Exit for B state
    // ExitB(){
    //     this.output += ' B-> ';
    //     assert(this.sm.Transition(this.a) == false);
    // }

    Test() {
        assert(this.sm.IsIn(null));

        this.sm.Transition(this.a);
        assert(this.sm.IsIn(this.a));

        this.sm.Transition(this.b);
        assert(this.sm.IsIn(this.b));

        this.sm.Transition(null);
        assert(this.sm.IsIn(null));

        assert(this.output == ' ->A  A->  ->B ');
    }    
}

foo = new Foo();
foo.Test();


